service/history/api/updateworkflowoptions/api.go

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package updateworkflowoptions
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"google.golang.org/protobuf/proto"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+	workflowpb "go.temporal.io/api/workflow/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/service/history/api"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func Invoke(
+	ctx context.Context,
+	request *historyservice.UpdateWorkflowExecutionOptionsRequest,
+	shardCtx shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+) (*historyservice.UpdateWorkflowExecutionOptionsResponse, error) {
+	ns, err := api.GetActiveNamespace(shardCtx, namespace.ID(request.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+	req := request.GetUpdateRequest()
+	ret := &historyservice.UpdateWorkflowExecutionOptionsResponse{}
+
+	err = api.GetAndUpdateWorkflowWithNew(
+		ctx,
+		nil,
+		definition.NewWorkflowKey(
+			ns.ID().String(),
+			req.GetWorkflowExecution().GetWorkflowId(),
+			req.GetWorkflowExecution().GetRunId(),
+		),
+		func(workflowLease api.WorkflowLease) (_ *api.UpdateWorkflowAction, updateError error) {
+			mutableState := workflowLease.GetMutableState()
+			defer func() { workflowLease.GetReleaseFn()(updateError) }()
+
+			if !mutableState.IsWorkflowExecutionRunning() {
+				// in-memory mutable state is still clean, let updateError=nil in the defer func()
+				// to prevent clearing and reloading mutable state while releasing the lock
+				return nil, consts.ErrWorkflowCompleted
+			}
+
+			// TODO (carly): dedupe by requestID
+			_ = req.GetRequestId()
```

Comment:
It is idempotent operation, right? Why do you need to dedup it?

Reply:
yes, it is idempotent.. you're right. I just saw other things like signal being deduped and wanted to make sure I did this too if it was necessary.

but I think it's not necessary because if there is no change, I don't add an event, so duplicate requests will not cause history size to increase.

Reply:
removed that line and todo